File: scripts/lib/epitome-eval-v3/trial-translate.ts
/**
 * Trial translation tester for Epitome Vol 3 evaluation
 * Sends sample paragraphs to DeepSeek and checks for gibberish
 */

import * as fs from 'fs';
import * as path from 'path';
import OpenAI from 'openai';

const CLEAN_DIR = path.join(__dirname, '../../../data/processed/epitome-of-histories-clean');

interface Paragraph {
  index: number;
  text: string;
}

interface ChapterData {
  chapterNumber: number;
  sourceContent: {
    paragraphs: Paragraph[];
  };
}

function getRandomParagraphs(n: number, seed: number): { chapter: number; index: number; text: string }[] {
  // Simple seeded random
  let s = seed;
  const random = () => {
    s = (s * 1103515245 + 12345) & 0x7fffffff;
    return s / 0x7fffffff;
  };

  const allParagraphs: { chapter: number; index: number; text: string }[] = [];

  for (let ch = 13; ch <= 18; ch++) {
    const filepath = path.join(CLEAN_DIR, `chapter-${ch.toString().padStart(3, '0')}.json`);
    if (!fs.existsSync(filepath)) continue;

    const data: ChapterData = JSON.parse(fs.readFileSync(filepath, 'utf-8'));
    for (const para of data.sourceContent.paragraphs) {
      allParagraphs.push({
        chapter: ch,
        index: para.index,
        text: para.text,
      });
    }
  }

  // Shuffle and take n
  for (let i = allParagraphs.length - 1; i > 0; i--) {
    const j = Math.floor(random() * (i + 1));
    [allParagraphs[i], allParagraphs[j]] = [allParagraphs[j], allParagraphs[i]];
  }

  return allParagraphs.slice(0, n);
}

async function translateText(client: OpenAI, text: string): Promise<string> {
  const response = await client.chat.completions.create({
    model: 'deepseek-chat',
    messages: [{
      role: 'user',
      content: `Translate this Byzantine Greek text to English. Provide only the translation, no explanations:\n\n${text}`,
    }],
    max_tokens: 2000,
    temperature: 0.3,
  });
  return response.choices[0]?.message?.content || '';
}

function checkTranslationQuality(translation: string): { hasIssues: boolean; issues: string[] } {
  const issues: string[] = [];

  // Check for untranslated sigla
  const siglaPatterns = ['Rwp', 'Dwp', 'RwpJDi', 'CDi', 'Owp', 'Bwp', 'wwp', 'AEDwwp'];
  for (const sig of siglaPatterns) {
    if (translation.toLowerCase().includes(sig.toLowerCase())) {
      issues.push(`Untranslated sigla: ${sig}`);
    }
  }

  // Check for Latin editorial terms translated literally
  const latinTerms = ['subscript', 'omisso', 'omissis', 'constanter', 'transponit', 'typotheta'];
  for (const term of latinTerms) {
    if (translation.toLowerCase().includes(term.toLowerCase())) {
      issues.push(`Latin apparatus term: ${term}`);
    }
  }

  // Check for obvious gibberish markers
  const gibberishMarkers = ['[?]', '???', 'unintelligible', 'unclear', 'corrupted'];
  for (const marker of gibberishMarkers) {
    if (translation.toLowerCase().includes(marker.toLowerCase())) {
      issues.push(`Gibberish marker: ${marker}`);
    }
  }

  return {
    hasIssues: issues.length > 0,
    issues,
  };
}

async function main() {
  const apiKey = process.env.DEEPSEEK_API_KEY;
  if (!apiKey) {
    console.error('Error: DEEPSEEK_API_KEY not set');
    process.exit(1);
  }

  const n = parseInt(process.argv[2] || '5', 10);
  const seed = parseInt(process.argv[3] || '42', 10);

  const client = new OpenAI({
    baseURL: 'https://api.deepseek.com',
    apiKey,
  });

  const paragraphs = getRandomParagraphs(n, seed);

  console.log(`=== TRIAL TRANSLATION TEST (${n} paragraphs, seed=${seed}) ===\n`);

  let cleanCount = 0;
  let problematicCount = 0;

  for (let i = 0; i < paragraphs.length; i++) {
    const para = paragraphs[i];
    console.log(`--- Test ${i + 1}: Chapter ${para.chapter}, Para ${para.index} ---`);
    console.log(`Source: ${para.text.slice(0, 200)}...`);

    try {
      const translation = await translateText(client, para.text.slice(0, 1500));
      console.log(`Translation: ${translation.slice(0, 300)}...`);

      const quality = checkTranslationQuality(translation);
      console.log(`Quality: ${quality.hasIssues ? 'PROBLEMATIC' : 'CLEAN'}`);
      if (quality.issues.length > 0) {
        console.log(`Issues: ${quality.issues.join(', ')}`);
        problematicCount++;
      } else {
        cleanCount++;
      }
    } catch (error: any) {
      console.log(`Error: ${error.message}`);
      problematicCount++;
    }

    console.log();

    // Small delay between requests
    await new Promise(r => setTimeout(r, 1000));
  }

  // Summary
  const cleanRate = (cleanCount / paragraphs.length) * 100;
  console.log(`\n=== SUMMARY ===`);
  console.log(`Total tested: ${paragraphs.length}`);
  console.log(`Clean translations: ${cleanCount} (${cleanRate.toFixed(1)}%)`);
  console.log(`Problematic: ${problematicCount}`);

  process.exit(problematicCount > 0 ? 1 : 0);
}

main().catch(err => {
  console.error(err);
  process.exit(1);
});

File: scripts/lib/epitome-eval-v3/trial_translate.py
#!/usr/bin/env python3
"""
Trial translation tester - sends sample paragraphs to DeepSeek API
and checks for gibberish/contamination in translation output.
"""

import json
import os
import random
import sys
from pathlib import Path

try:
    import openai
except ImportError:
    print("Error: openai package not installed. Run: pip install openai", file=sys.stderr)
    sys.exit(1)

def get_random_paragraphs(n: int = 5, seed: int = 42) -> list:
    """Get n random paragraphs from the cleaned files."""
    random.seed(seed)
    base_dir = Path('/Users/bryancheong/claude_projects/translation-wiki/data/processed/epitome-of-histories-clean')

    all_paragraphs = []
    for chapter_num in range(13, 19):
        filepath = base_dir / f'chapter-{chapter_num:03d}.json'
        if not filepath.exists():
            continue

        with open(filepath, 'r', encoding='utf-8') as f:
            data = json.load(f)

        paragraphs = data.get('sourceContent', {}).get('paragraphs', [])
        for para in paragraphs:
            all_paragraphs.append({
                'chapter': chapter_num,
                'index': para.get('index', -1),
                'text': para.get('text', ''),
            })

    return random.sample(all_paragraphs, min(n, len(all_paragraphs)))

def translate_text(client, text: str) -> str:
    """Translate Byzantine Greek to English using DeepSeek."""
    response = client.chat.completions.create(
        model="deepseek-chat",
        messages=[{
            "role": "user",
            "content": f"Translate this Byzantine Greek text to English. Provide only the translation, no explanations:\n\n{text}"
        }],
        max_tokens=2000,
        temperature=0.3,
    )
    return response.choices[0].message.content

def check_translation_quality(translation: str) -> dict:
    """Check translation for signs of contamination/gibberish."""
    issues = []

    # Check for untranslated sigla
    sigla_patterns = ['Rwp', 'Dwp', 'RwpJDi', 'CDi', 'Owp', 'Bwp']
    for sig in sigla_patterns:
        if sig.lower() in translation.lower():
            issues.append(f"Untranslated sigla: {sig}")

    # Check for Latin editorial terms translated literally
    latin_terms = ['subscript', 'omisso', 'constanter', 'transponit', 'typotheta']
    for term in latin_terms:
        if term.lower() in translation.lower():
            issues.append(f"Latin apparatus term: {term}")

    # Check for obvious gibberish markers
    gibberish_markers = ['[?]', '???', 'unintelligible', 'unclear']
    for marker in gibberish_markers:
        if marker.lower() in translation.lower():
            issues.append(f"Gibberish marker: {marker}")

    # Check for page marker artifacts
    if any(f"page {i}" in translation.lower() and "manuscript" not in translation.lower()
           for i in range(300, 500)):
        issues.append("Possible page marker artifact")

    return {
        'has_issues': len(issues) > 0,
        'issues': issues,
        'confidence': 'LOW' if len(issues) > 2 else ('MEDIUM' if len(issues) > 0 else 'HIGH'),
    }

def main():
    api_key = os.environ.get('DEEPSEEK_API_KEY')
    if not api_key:
        print("Error: DEEPSEEK_API_KEY environment variable not set", file=sys.stderr)
        sys.exit(1)

    n = int(sys.argv[1]) if len(sys.argv) > 1 else 5
    seed = int(sys.argv[2]) if len(sys.argv) > 2 else 42

    client = openai.OpenAI(
        base_url="https://api.deepseek.com",
        api_key=api_key,
    )

    paragraphs = get_random_paragraphs(n, seed)

    print(f"=== TRIAL TRANSLATION TEST ({n} paragraphs, seed={seed}) ===\n")

    results = {
        'total': len(paragraphs),
        'clean_translations': 0,
        'problematic_translations': 0,
        'details': [],
    }

    for i, para in enumerate(paragraphs, 1):
        print(f"--- Test {i}: Chapter {para['chapter']}, Para {para['index']} ---")
        print(f"Source: {para['text'][:200]}...")

        try:
            translation = translate_text(client, para['text'][:1500])  # Limit length
            print(f"Translation: {translation[:300]}...")

            quality = check_translation_quality(translation)
            print(f"Quality: {quality['confidence']}")
            if quality['issues']:
                print(f"Issues: {', '.join(quality['issues'])}")
                results['problematic_translations'] += 1
            else:
                results['clean_translations'] += 1

            results['details'].append({
                'chapter': para['chapter'],
                'index': para['index'],
                'quality': quality,
            })

        except Exception as e:
            print(f"Error: {e}")
            results['problematic_translations'] += 1

        print()

    # Summary
    clean_rate = results['clean_translations'] / results['total'] * 100 if results['total'] > 0 else 0
    print(f"\n=== SUMMARY ===")
    print(f"Total tested: {results['total']}")
    print(f"Clean translations: {results['clean_translations']} ({clean_rate:.1f}%)")
    print(f"Problematic: {results['problematic_translations']}")

    return results['problematic_translations'] == 0

if __name__ == '__main__':
    success = main()
    sys.exit(0 if success else 1)
